functions

```diff
@@ -397,24 +386,47 @@ add_module() {
                 done
                 ;;
         esac
-    done < <(modinfo -b "$_optmoduleroot" -k "$KERNELVERSION" -0 "$target" 2>/dev/null)
+    done < <(modinfo -b "$_optmoduleroot" -k "$KERNELVERSION" -0 "${@%\?}" 2>/dev/null)
 
-    if (( !found )); then
-        (( ign_errors || _addedmodules["$target"] )) && return 0
-        error "module not found: \`%s'" "$target"
-        return 1
-    fi
+    for m in "${@//-/_}"; do
+        if [[ $m = *\? ]]; then
+            continue
+        fi
+
+        if (( _addedmodules[$m] )); then
+            continue
+        fi
+
+        warning "module not found: %s" "$m"
+    done
 
-    if (( ${#firmware[*]} )); then
-        add_firmware "${firmware[@]}" ||
-            warning 'Possibly missing firmware for module: %s' "$target"
+    if (( ! ${#_add_modules[@]} )); then
+        return
     fi
 
-    # handle module quirks
-    case $target in
+    _add_module "${_add_modules[@]}"
+}
+
+# add_module adds a kernel modules $@ and all it's dependencies and firmware
+# to the image. ?-suffix ignores errors.
+add_module() {
+    local target mod
+
+    [[ $KERNELVERSION == none ]] && return 0
+
+    for target; do
+        target=${target%.ko*}
+        target=${target//-/_}
+
+        mod=${target%\?}
+
+        (( _addmodule[$mod] )) && continue
+        _addmodule[$mod]=1
```

Comment:
I think I understand why you have this, but it took me a solid 5-10 minutes to figure it out because the naming lead me to believe that it was a typo. Please drop it, or if you can show significant benetfit, add a comment explaining why we want the second layer of caching.

Reply:
The amount of modules modinfo needs to resolve with this change drops in half and reduces the time spent in modinfo significantly.
```sh
# without
$ time ./mkinitcpio -c /dev/null -A block,filesystems,keyboard &> /dev/null

real	0m4.062s
user	0m3.873s
sys	0m1.566s

# with
$ time ./mkinitcpio -c /dev/null -A block,filesystems,keyboard &> /dev/null

real	0m2.423s
user	0m2.214s
sys	0m0.871s
```